include/unifex/find_if.hpp

```diff
@@ -24,63 +26,219 @@
+    template <typename Iterator, typename... Values>
+    void set_value(std::tuple<Iterator, Values...>&& packedResult) && noexcept {
+      operation_state_.cleanup();
       try {
-        auto result = find_if_helper(begin_it, end_it, values...);
-
-        unifex::set_value((Receiver &&) receiver_, std::move(result), (Values &&) values...);
-      } catch (...) {
-        unifex::set_error((Receiver &&) receiver_, std::current_exception());
+        unpack_helper(
+          (OutputReceiver&&) output_receiver_,
+          std::move(packedResult),
+          std::make_index_sequence<std::tuple_size_v<std::tuple<Iterator, Values...>>>{});
+      } catch(...) {
+        unifex::set_error((OutputReceiver&&)output_receiver_, std::current_exception());
       }
     }
-  }
+
+    template <typename Error>
+    void set_error(Error&& error) && noexcept {
+      operation_state_.cleanup();
+      unifex::set_error((OutputReceiver &&) output_receiver_, (Error &&) error);
+    }
+
+    void set_done() && noexcept {
+      operation_state_.cleanup();
+      unifex::set_done((OutputReceiver &&) output_receiver_);
+    }
+
+    template(typename CPO, typename R)
+        (requires is_receiver_query_cpo_v<CPO> AND same_as<R, unpack_receiver<OutputReceiver>>)
+    friend auto tag_invoke(CPO cpo, const R& r) noexcept(
+        is_nothrow_callable_v<CPO, const OutputReceiver&>)
+        -> callable_result_t<CPO, const OutputReceiver&> {
+      return std::move(cpo)(std::as_const(r.output_receiver_));
+    }
+  };
+
+  struct find_if_helper {
+    Func func_;
+
+    template<typename Iterator, typename... Values>
+    auto operator()(
+        const unpack_receiver<Receiver>& /*unused*/,
+        const sequenced_policy&,
+        Iterator begin_it,
+        Iterator end_it,
+        Values&&... values) noexcept {
+      // Sequential implementation
+      return unifex::transform(
+          unifex::just(std::forward<Values>(values)...),
+          [this, begin_it, end_it](auto... values) {
+            for(auto it = begin_it; it != end_it; ++it) {
+              if(std::invoke((Func &&) func_, *it, values...)) {
+                return std::tuple<Iterator, Values...>(it, std::move(values)...);
+              }
+            }
+            return std::tuple<Iterator, Values...>(end_it, std::move(values)...);
+          }
+        );
+    }
+
+    // Cancellable parallel algorithm.
+    // This version is two phase to avoid a non-trivial atomic in the middle.
+    // With more built-in algorithms it can be simplified:
+    //  * let_with to allocate non-movable state in the operation state.
+    //  * unpack to deal with tuple to pack conversion
+    // It could also be simplified by making more of the code custom, but I wanted
+    // to demonstrate reuse of internal algorithms to build something more compelex
+    // and cancellable.
+    template<typename Iterator, typename... Values>
+    auto operator()(
+        const unpack_receiver<Receiver>& receiver,
```

Comment:
Seems unusual to pass in the unpack_receiver when its only used for the scheduler.
Could we just pass in the scheduler here instead?

Reply:
Yes, although as it is unused in the other overload there's marginal inefficiency there.